### ISSUE TITLE
[css-color-6] It should be possible to specify the contrast algorithm for color-contrast()

### ISSUE THREAD

**Issue by LeaVerou** [w3c/csswg-drafts#7356]
Since we know that WCAG 2.1 has [huge known problems](https://www.cedc.tools/article.html) and WCAG 3.0 isn't quite ready yet, we should not ship a `color-contrast()` that mandates use of a broken algorithm.
Instead, the algorithm should be explicitly specified with either syntax.

Since passing levels depend on the algorithm and 4.5 in WCAG 2.1 may not be the same as 4.5 in another algorithm, I'd suggest it's specified via functions, e.g. `color-contrast(white vs yellow, blue to wcag2(4.5))` or `color-contrast(white vs yellow, blue to wcag2(AA))`

*(Issue filed following breakout discussions between @svgeesus, @fantasai, @argyleink and myself)*

**Comment by SebastianZ** [comment]
Question is also what to do if the specified algorithm (or maybe even the contrast ratio) are not supported by the UA. Should it be invalid at computed value time or should the UA fall back to its default algorithm or something else?

Sebastian

**Comment by argyleink** [comment]
works for me. 
i'm only a tiny bit curious why it's a function, not really how other css functions work, but i mean it matches the mental model so thats nice.

**Comment by Myndex** [comment]
HI Lea @LeaVerou 

> .._.levels don't mean anything outside their algorithm, ....  (e.g. `wcag-aa`),.._

Not only that, but: **AA** encompasses two discrete ratios, 3:1 and 4.5:1, and **AAA** encompasses three distinct ratios, those of AA, and the additional 7:1.

And the 3:1, 4.5:1, and 7:1  as calculated by WCAG 2 math are _different than every other contrast math,_ and the actual results skew differently relative to the overall total luminances being tested as does every other contrast math or method, ever...

And I know you know this but for anyone reading along at home, among the reasons is that there is no such thing as an absolute contrast: **Contrast is not real,** just as color is not real, it is a context sensitive perception and about as hard to nail down as grasping a wet bar of soap in the middle of a typhoon while being chased by hungry wild geese.

**Comment by svgeesus** [comment]
> Not only that, but: AA encompasses two discrete ratios, 3:1 and 4.5:1, and AAA encompasses three distinct ratios, those of AA, and the additional 7:1.

Yes, we know; in CSS [we have separate keywords for those](https://drafts.csswg.org/css-color-6/#colorcontrast):

>> The keyword AA is equivalent to 4.5, AA-large is equivalent to 3, AAA is equivalent to 7, and AAA-large is equivalent to 4.5 .

> And the 3:1, 4.5:1, and 7:1 as calculated by WCAG 2 math are different than every other contrast math,

Yes, that is what "Because levels don't mean anything outside their algorithm" is saying.

**Comment by yatil** [comment]
> Yes, we know; in CSS [we have separate keywords for those](https://drafts.csswg.org/css-color-6/#colorcontrast):
> 
> 
> >> The keyword AA is equivalent to 4.5, AA-large is equivalent to 3, AAA is equivalent to 7, and AAA-large is equivalent to 4.5 .

I would be cautious to have the `AA`, `AA-large`, `AAA` and `AAA-large` keywords as they only apply to text. For graphical elements AA requires only a 3:1 contrast.

While I think it would be neat to be able to specify the contrast by level, it might be better to – by default – assume 4.5:1, and allow users to specify a different contrast ratio if they want to. A 4.5:1 contrast ratio is sufficient or better for all cases apart from AAA (non-large), and there are not many who aspire to AAA (unfortunately).

(I think it would also be good to explore requiring the whole contrast ratio `X.x:1` instead of just the number before the `:1`. I floated the idea of leaving off the `:1` a few years ago and while details escape me, having a ratio was seen as communicating clearer that this is about how two colors behave in coordination with each other. In the end, it might be good to align with WCAG/AG WG on that matter.)

**Comment by Myndex** [comment]
> _...A 4.5:1 contrast ratio is sufficient or better for all cases apart..._ 

In terms of **_actual_** accessibility, 4.5:1 is grossly insufficient for a vast portion of the visual range, most especially dark colors and "dark mode". (And oddly, 4.5:1 is more than needed for some other use cases and low spatial frequency elements).

As I've mentioned elsewhere, a numerical value can be more useful. Perhaps a single default (and if it is WCAG 2 that default should arguably be 7:1) but rather that accept only a few levels, a numerical value is beneficial. In the case of the WCAG 2 algorithm, that would allow a designer to specify 10:1, which is a more ideal level for body text.

As I mentioned in [this post in 7359 ](https://github.com/w3c/csswg-drafts/issues/7359#issuecomment-1208744612) combining an identifier with a contrast value would allow unambiguous contrast settings.
